src/Common/tests/System/Net/Http/Http2LoopbackServer.cs

```diff
@@ -301,20 +540,89 @@ public async Task SendDefaultResponseAsync(int streamId)
+
+            if (headers != null)
+            {
+                foreach (HttpHeaderData headerData in headers)
+                {
+                    bytesGenerated += EncodeHeader(headerData, headerBlock.AsSpan().Slice(bytesGenerated));
+                }
+            }
+
+            FrameFlags flags = FrameFlags.EndHeaders;
+            if (endStream)
+            {
+                flags |= FrameFlags.EndStream;
+            }
+
+            HeadersFrame headersFrame = new HeadersFrame(headerBlock.AsMemory().Slice(0, bytesGenerated), flags, 0, 0, 0, streamId);
+            await WriteFrameAsync(headersFrame).ConfigureAwait(false);
+        }
+
+        public async Task SendResponseDataAsync(int streamId, ReadOnlyMemory<byte> responseData, bool endStream)
         {
-            DataFrame dataFrame = new DataFrame(data, endStream ? FrameFlags.EndStream : FrameFlags.None, 0, streamId);
+            DataFrame dataFrame = new DataFrame(responseData, endStream ? FrameFlags.EndStream : FrameFlags.None, 0, streamId);
             await WriteFrameAsync(dataFrame).ConfigureAwait(false);
         }
 
-        public void Dispose()
+        public async Task SendResponseBodyAsync(int streamId, ReadOnlyMemory<byte> responseBody)
+        {
+            // Only support response body if it fits in a single frame, for now
+            // In the future we should separate the body into chunks as needed,
+            // and if it's larger than the default window size, we will need to process window updates as well.
+            if (responseBody.Length > Frame.MaxFrameLength)
+            {
+                throw new Exception("Response body too long");
+            }
+
+            await SendResponseDataAsync(streamId, responseBody, true);
+        }
+
+        public override void Dispose()
         {
             if (_listenSocket != null)
             {
                 _listenSocket.Dispose();
                 _listenSocket = null;
             }
         }
+
+        //
+        // GenericLoopbackServer implementation
+        //
+
+        public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode statusCode = HttpStatusCode.OK, IList<HttpHeaderData> headers = null, string content = null)
+        {
+            await EstablishConnectionAsync();
+
+            (int streamId, HttpRequestData requestData) = await ReadAndParseRequestHeaderAsync();
+
+            // We are about to close the connection, after we send the response.
+            // So, send a GOAWAY frame now so the client won't inadvertantly try to reuse the connection.
+            await SendGoAway(streamId);
+
+            if (content == null)
+            {
+                await SendResponseHeadersAsync(streamId, true, statusCode, headers);
+            }
+            else
+            {
+                await SendResponseHeadersAsync(streamId, false, statusCode, headers);
+                await SendResponseBodyAsync(streamId, Encoding.ASCII.GetBytes(content));
```

Comment:
Why ASCII here?

Reply:
No particularly reason. It could be UTF8 or whatever. For the cases where it's used today, it doesn't really matter.

What we probably need in the future is a way to specify arbitrary content as a ReadOnlySpan<byte>.

Reply:
That seems reasonable. I'll resolve this.